### PR TITLE
[exporters/prometheus] Handle attribute key collisions from sanitation

### LINKED ISSUE #2290: Prometheus exporter: Handle attribute key collisions from sanitization
**Is your feature request related to a problem?**

From https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#metric-attributes,

> OpenTelemetry Metric Attributes MUST be converted to [Prometheus labels](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels). String Attribute values are converted directly to Metric Attributes, and non-string Attribute values MUST be converted to string attributes following the [attribute specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute). Prometheus metric label keys are required to match the following regex: `[a-zA-Z_]([a-zA-Z0-9_])*`. Metrics from OpenTelemetry with unsupported Attribute names MUST replace invalid characters with the _ character. Multiple consecutive _ characters MUST be replaced with a single _ character. This may cause ambiguity in scenarios where multiple similar-named attributes share invalid characters at the same location. **In such unlikely cases, if multiple key-value pairs are converted to have the same Prometheus key, the values MUST be concatenated together, separated by ;, and ordered by the lexicographical order of the original keys.**

**Describe the solution you'd like**

One solution is to group colliding keys, and then join them with `;`, like the go exporter does:
https://github.com/open-telemetry/opentelemetry-go/blob/02616a25c68e674a04df01a1b7ba185a2db8cd5a/exporters/prometheus/exporter.go#L284-L305

Another solution is to sort the keys first, and then check whether it collides with the key from the previous iteration, like i'm working on in java: https://github.com/open-telemetry/opentelemetry-java/pull/5717

### PR DESCRIPTION
**Pull request by punya**
Fixes #2290

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

*   [x] `CHANGELOG.md` updated for non-trivial changes
*   [x] Unit tests have been added
*   [x] Changes in public API reviewed

### CODE REVIEW

File: exporters/prometheus/src/exporter_utils.cc [hunk 1]
```diff
@@ -258,16 +258,37 @@ void PrometheusExporterUtils::SetData(std::vector<T> values,
...
+  if (labels.empty())
   {
-    metric.label.resize(labels.size());
-    size_t i = 0;
-    for (auto const &label : labels)
+    return;
+  }
+
+  // Concatenate values for keys that collide after sanitation.
+  // Note that attribute keys are sorted, but sanitized keys can be out-of-order.
```
Comment (lalitb) [inline exporters/prometheus/src/exporter_utils.cc:267]: The SDK doesn't sort the attribute keys. Unless the instrumentation library is recording them in sorted order, I am not sure if below logic would work.

Comment (punya) [reply exporters/prometheus/src/exporter_utils.cc:267]: The type is `OrderedAttributeMap` which is an alias for `std::vector<std::string, ...>`. I see that a similar unordered type exists, but that's not the one used here.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks.

**Review by marcalff — APPROVED** [review]
LGTM, thanks for the fix.

**Pull request merged**